cheatsheets/Protect_FileUpload_Against_Malicious_File.md

```diff
@@ -1,597 +1,138 @@
-                                throw new IOException("Format of the original image is not" + 
-                                                      " supported for write operation !");
-                            }
-
-                        }
-                        imageParser.writeImage(sanitizedImage, fos, new HashMap<>());
-                    }
-
-                }
-
-                // Set state flag
-                safeState = true;
-            }
-        } catch (Exception e) {
-            safeState = false;
-            LOG.warn("Error during Image file processing !", e);
-        }
-
-        return safeState;
-    }
-}
+```python
+# This code snippet only serves as an example, and should not be used
+blacklist = ["php", "php5"]  # can be replaced with a regex pattern, but that would be even weaker, such as \.php$
+for bad in blacklist:
+    if "file.php".endswith(bad):
+        // raise exception
 ```
 
-# Sources of the prototype
+### User Permissions
+
+Before any file upload service is accessed, proper validation should occur on two levels for the user uploading a file:
+
+- Authentication level
+  - The user should be a registered user, or an identifiable user, in order to set restrictions and limitations for their upload capabilities
+- Authorization level
+  - Make sure that the user's accesses allow them to upload a file, or multiple files for administrative cases.
+
+### Content-Type Validation
+
+Other than defining the extension of the uploaded file, its MIME-type can be checked for a quick protection against simple file upload attacks.
+
+This can be done preferrably in a whitelist approach; otherwise, this can be done in a blacklist approach.
+
+> This should not be used on its own, as bypassing it is pretty common and easy.
+
+### File Name Sanitization
+
+Multiple threats relate to the file-name, thus, it is essential to ensure that the file-name is stored in a secure fashion.
+
+This can be done by creating a **random string** as a file-name. If the file-name is required by the business needs, proper input validation should be done for client-side (_e.g._ active content) and back-end side (_e.g._ special files overwrite or creation) attack vectors. File-name length limits should be taken into consideration based on the system storing the files, as each system has its own file name length limit.
+
+### File Storage Location
+
+All file uploaded will be stored in a location on the server. This location must be chosen based on security and business requirements. The following points are set by security priority, and are inclusive:
+
+1. Store the files on a **different host**, which allow for complete segragation of duties between the application serving the user, and the host handling file uploads and their storage.
+2. Store the files in a **private directory**, where only administrative access is allowed.
+3. Store the files on a **public directory**, and set them in write permissions only.
+
```

Comment:
There's also the option to store the files in a (No)SQL database as BLOBs. This can resolve lots of filename/path related issues, but also makes it harder to AV scan then.

Reply:
They can be stored in DBs if they aren't that big and the databases can be properly scaled. I'll add a point to it

Reply:
This is the kind of thing that a NoSQL database (essentially a key:value store for arbitrary data) could be perfect for.